fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java

```diff
@@ -2089,6 +2093,210 @@ private boolean isStateCacheableInternal() {
         return recordStoreStateRef.get().getStoreHeader().getCacheable();
     }
 
+    private void validateCanAccessHeaderUserFields() {
+        if (formatVersion < HEADER_USER_FIELDS_FORMAT_VERSION) {
+            throw new RecordCoreException("cannot access header user fields at current format version",
+                    LogMessageKeys.FORMAT_VERSION, formatVersion);
+        }
+    }
+
+    /**
+     * Get one of the value of a user-settable field from the store header. Each of these fields are written into
+     * this record store's store header. This is loaded automatically by the record store as part of
+     * {@link Builder#createOrOpenAsync()} or one of its variants. This means that reading this
+     * information from the header does not require any additional communication with database assuming that
+     * the record store has already been initialized. As a result, this is not a blocking call.
+     *
+     * <p>
+     * Using this feature also requires that the record store be on format version {@link #HEADER_USER_FIELDS_FORMAT_VERSION}
+     * or higher. There is no change to the on-disk format version from the previous format version except that
+     * extra fields in the store header were added. No data migration is necessary to upgrade to that version if
+     * an existing store is on the previous format version, but the new version is used to prevent the user from serializing
+     * data and then not being able to read it from instances running an older version of the Record Layer.
+     * </p>
+     *
+     * @param userField the name of the user-settable field to read
+     * @return the value of that field in the header or {@code null} if it is unset
+     * @see #setHeaderUserFieldAsync(String, ByteString)
+     */
+    @Nullable
+    public ByteString getHeaderUserField(@Nonnull String userField) {
+        validateCanAccessHeaderUserFields();
+        if (recordStoreStateRef.get() == null) {
+            throw new RecordCoreException("cannot get field from header as record store state is null");
```

Comment:
For this sort of exception, it there any guidance we could give in the text as to what this means or what could be done with it?  Or, at least, if the store state being `null` is a unexpected state to be found in?   Also, maybe this should include logging info of the subspace provider of the store.